### ISSUE TITLE
add a nomemory_allocator to the _testcapi module

### ISSUE THREAD

**Issue by 490c593f-f636-409f-bb35-6abeb38a4595** [python/cpython#74880]
BPO | [30695](https://bugs.python.org/issue30695)
--- | :---
Nosy | @vstinner, @xdegaye, @serhiy-storchaka
PRs | <li>python/cpython#2406</li><li>python/cpython#4083</li>
Files | <li>[nomemory_allocator.patch](https://bugs.python.org/file46958/nomemory_allocator.patch "Uploaded as text/plain at 2017-06-18.10:21:10 by @xdegaye")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-10-23.13:24:32.179>
created_at = <Date 2017-06-18.10:21:11.337>
labels = ['3.7', 'type-feature', 'tests']
title = 'add a nomemory_allocator to the _testcapi module'
updated_at = <Date 2017-10-23.13:24:32.179>
user = 'https://github.com/xdegaye'
```

bugs.python.org fields:
```python
activity = <Date 2017-10-23.13:24:32.179>
actor = 'xdegaye'
assignee = 'none'
closed = True
closed_date = <Date 2017-10-23.13:24:32.179>
closer = 'xdegaye'
components = ['Tests']
creation = <Date 2017-06-18.10:21:11.337>
creator = 'xdegaye'
dependencies = []
files = ['46958']
hgrepos = []
issue_num = 30695
keywords = ['patch']
message_count = 19.0
messages = ['296266', '296276', '296299', '296318', '296320', '296323', '296331', '296333', '296338', '296339', '296340', '296889', '296893', '296894', '296900', '297412', '297481', '304797', '304802']
nosy_count = 3.0
nosy_names = ['vstinner', 'xdegaye', 'serhiy.storchaka']
pr_nums = ['2406', '4083']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue30695'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
Add the set_nomemory_allocator() function to \_testcapi that sets a no memory allocator.

**Comment by serhiy-storchaka** [comment]
I think the allocators must just return NULL, without setting the error.

**Comment by vstinner** [comment]
Nice. Do you know my https://pypi.python.org/pypi/pyfailmalloc project? It
helped me to identify and fix dozens of code which didn't handle properly
memory allocation failures. Seach for "pyfailmalloc" in the bug tracker
(closed issues) ;-)

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
Yes, actually I thought first about re-opening bpo-19817 instead of opening this new issue.

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
> I think the allocators must just return NULL, without setting the error.
You are right, thanks.

**Comment by vstinner** [comment]
The main difference between your change and pyfailmalloc is the ability in pyfailmalloc to only fail in N allocations. If you want to test various code paths, you need to allow N allocations to reach the deep code that you want to test.

My code is something like 100 lines of C code.
https://bitbucket.org/haypo/pyfailmalloc/src/cdaf3609a30b88243d04e79d6074f3b28d9b64e3/failmalloc.c?at=default&fileviewer=file-view-default

IMHO it's small enough to fit directly into \_testcapi.

The question is more what do you want to do with it? It was proposed to "include" pyfailmalloc directly in the Python test suite. Maybe add an option to enable it?

I ran the Python test suite with pyfailmalloc in gdb using this script:
https://bitbucket.org/haypo/pyfailmalloc/src/cdaf3609a30b88243d04e79d6074f3b28d9b64e3/debug_cpython.gdb?at=default&fileviewer=file-view-default

./python -m test -F -r -v -x test_tracemalloc

I waited for the next \*crash* and ignored all expected tests failing with MemoryError. --forever (-F) is nice to run random tests until you get a crash.

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
The chain of events following a call to set_nomemory_allocator() is deterministic, this is another difference with pyfailmalloc. Also what happens then after this call is not very close to real life as memory is never freed.

Having the possibility to trigger memory exhaustion in N allocations instead of rightaway would be very useful. Maybe this could be done by including pyfailmalloc in \_testcapi with some changes to enable multiple behaviors, among which one of them would be deterministic.

**Comment by vstinner** [comment]
Your code can be reproduced with pyfailmalloc using N=0 :-)

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
I am not sure, failmalloc.enable(range) accepts only range \> 1, hook_malloc() fails only once instead of permanently and hook_free() does free memory.

**Comment by vstinner** [comment]
> I am not sure, failmalloc.enable(range) accepts only range \> 1, hook_malloc() fails only once instead of permanently and hook_free() does free memory.

All these things can change :-)

> Also what happens then after this call is not very close to real life as memory is never freed.

Why not freeing memory?

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
Good point. Not freeing the memory made the implementation of set_nomemory_allocator() simpler, no need to call PyMem_GetAllocator(). Forget about this point, sorry :-)

**Comment by vstinner** [comment]
I prefer your new PR. But is it useful to test N memory allocation failures in a row? The API and the code would be simpler if we would only test a single failure. What do you think?

I know that pyfailmalloc is different, but I'm not sure that pyfailmalloc design is right neither :-)

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
For the record, while working on the test case of [PR 2406](https://github.com/python/cpython/pull/2406), I found by chance that the following script:

```py
# Script start.
import _testcapi
class C(): pass
_testcapi.set_nomemory(0, 5)
C()
# Script end.
```


fails with:

python: [Objects/call.c:89](https://github.com/python/cpython/blob/main/Objects/call.c#L89): \_PyObject_FastCallDict: Assertion \`!PyErr_Occurred()' failed.
Aborted (core dumped)

I will create a new issue when the current issue is closed.

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
> But is it useful to test N memory allocation failures in a row?

I think it is useful. See my previous post.

**Comment by vstinner** [comment]
2017-06-26 15:34 GMT+02:00 Xavier de Gaye \<report@bugs.python.org\>:
> python: [Objects/call.c:89](https://github.com/python/cpython/blob/main/Objects/call.c#L89): \_PyObject_FastCallDict: Assertion \`!PyErr_Occurred()' failed.
> Aborted (core dumped)
>
> I will create a new issue when the current issue is closed.

Oh, I'm curious about that one :-)

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
> Oh, I'm curious about that one :-)

This is bpo-30817.

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
New changeset 85f643023fed3d4e2fb8e399f9ad57f3a65ef237 by xdegaye in branch 'master':
bpo-30695: Add set_nomemory(start, stop) to \_testcapi (GH-2406)
https://github.com/python/cpython/commit/85f643023fed3d4e2fb8e399f9ad57f3a65ef237

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
Test cases in issues bpo-30697 and bpo-30817, back ported to 3.6, need \_testcapi.set_nomemory().

**Comment by 490c593f-f636-409f-bb35-6abeb38a4595** [comment]
New changeset aaf6a3dbbdb9754f98d480b468adfcae0f66e3a2 by xdegaye (Miss Islington (bot)) in branch '3.6':
[3.6] bpo-30695: Add set_nomemory(start, stop) to \_testcapi (GH-2406) (bpo-4083)
https://github.com/python/cpython/commit/aaf6a3dbbdb9754f98d480b468adfcae0f66e3a2
